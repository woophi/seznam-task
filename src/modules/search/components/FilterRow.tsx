import { NativeCheckBox } from 'modules/native-checkbox';
import { useFela } from 'react-fela';

type Props<T> = {
  title: string;
  options: {
    label: string;
    value: T;
    checked: boolean;
  }[];
  onChange: (value: T) => void;
};

export function FilterRow<T>({ title, options, onChange }: Props<T>) {
  const { css } = useFela();
  return (
    <div className={css({ display: 'flex', flexDirection: 'column', margin: '.25rem .5rem 0 1rem', cursor: 'pointer' })}>
      <b className={css({ marginBottom: '.5rem' })}>{title}</b>
      {options.map(option => (
        <span key={`${option.value}`} className={css({ marginBottom: '.5rem' })} onClick={() => onChange(option.value)}>
          <NativeCheckBox onChange={() => onChange(option.value)} checked={option.checked} computedId={`${option.value}`} />
          {option.label}
        </span>
      ))}
    </div>
  );
}
