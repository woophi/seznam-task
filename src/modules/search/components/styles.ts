import { FelaRule } from 'react-fela';

export const searchContainer: FelaRule<{ open: boolean }> = ({ open }) => ({
  margin: '1rem',
  borderRadius: '.25rem',
  boxShadow: '0 1px 2px 0 rgb(0 0 0 / 12%), 0 0 3px 0 rgb(0 0 0 / 6%)',
  display: 'flex',
  overflow: 'hidden',
  height: open ? '320px' : '48px',
  transition: 'ease-in-out .2s height',
  flexDirection: 'column'
});
export const searchInput: FelaRule = () => ({
  border: 'none',
  width: '100%',
  height: '48px',
  outline: 'none',
  lineHeight: '24px',
  fontSize: '21px',
  padding: 0
});
export const filterIcon: FelaRule = () => ({
  padding: '1rem',
  cursor: 'pointer',
  borderRadius: '.25rem',
  '&:hover': {
    backgroundColor: 'whitesmoke'
  }
});
