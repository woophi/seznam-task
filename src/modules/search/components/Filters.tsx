import { FiltersModel, MobileOS, MobileProcessor } from 'core/models';
import { memo } from 'react';
import { FilterRow } from './FilterRow';

type Props = {
  model: FiltersModel;
  changeOS: (v: MobileOS) => void;
  changeProcessor: (v: MobileProcessor) => void;
};

export const Filters = memo<Props>(({ model, changeProcessor, changeOS }) => {
  return (
    <>
      <FilterRow<MobileOS>
        onChange={changeOS}
        title="OS"
        options={[
          {
            checked: model.os.includes(MobileOS.Android),
            label: MobileOS.Android,
            value: MobileOS.Android
          },
          {
            checked: model.os.includes(MobileOS.IOS),
            label: MobileOS.IOS,
            value: MobileOS.IOS
          }
        ]}
      />
      <FilterRow<MobileProcessor>
        onChange={changeProcessor}
        title="Processor"
        options={[
          {
            checked: model.processor.includes(MobileProcessor.Apple13),
            label: MobileProcessor.Apple13,
            value: MobileProcessor.Apple13
          },
          {
            checked: model.processor.includes(MobileProcessor.Apple14),
            label: MobileProcessor.Apple14,
            value: MobileProcessor.Apple14
          },
          {
            checked: model.processor.includes(MobileProcessor.Exynos2100),
            label: MobileProcessor.Exynos2100,
            value: MobileProcessor.Exynos2100
          },
          {
            checked: model.processor.includes(MobileProcessor.Snapdragon888),
            label: MobileProcessor.Snapdragon888,
            value: MobileProcessor.Snapdragon888
          }
        ]}
      />
    </>
  );
});
