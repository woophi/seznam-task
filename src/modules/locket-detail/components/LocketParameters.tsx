import { objKeys } from 'core/data/utils';
import { LocketDetailParameters, LocketParameters as LocketParametersType } from 'core/models';
import { firstUppercase } from 'core/shared/styles';
import { memo } from 'react';
import { useFela } from 'react-fela';

type Props = {
  parameters: LocketDetailParameters | LocketParametersType;
};

const parameterPostfix = (key: keyof LocketDetailParameters | keyof LocketParametersType) => {
  switch (key) {
    case 'internalMemory':
    case 'ram':
      return ' GB';

    case 'display':
      return '"';

    default:
      return '';
  }
};

export const LocketParameters = memo<Props>(({ parameters }) => {
  const { css } = useFela();

  return (
    <div className={css({ marginTop: '1rem' })}>
      {objKeys(parameters).map(key => (
        <div key={key}>
          <b className={css(firstUppercase)}>{key}: </b>
          {parameters[key]}
          {parameterPostfix(key)}
        </div>
      ))}
    </div>
  );
});
