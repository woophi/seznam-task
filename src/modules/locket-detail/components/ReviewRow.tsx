import { faAngry, faMeh, faSmileBeam } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ReviewInfo } from 'core/models';
import { firstLetterUp } from 'core/shared/styles';
import { FC } from 'react';
import { useFela } from 'react-fela';

type Props = {
  info: ReviewInfo;
};

export const ReviewRow: FC<Props> = ({ info: { name, positive, feedback, created } }) => {
  const { css } = useFela();

  const icon = positive === null ? faMeh : positive ? faSmileBeam : faAngry;
  const iconColor = positive === null ? undefined : positive ? '#8cc643' : '#FF5A32';
  return (
    <div>
      <div className={css({ display: 'flex', alignItems: 'center' })}>
        <b className={css(firstLetterUp, { padding: '.25rem 0', marginRight: '.5rem' })}>{name}</b>{' '}
        <FontAwesomeIcon icon={icon} className={css({ color: iconColor, marginRight: '.5rem' })} />
        <div className={css({ opacity: 0.6, fontStyle: 'italic' })}>
          {created.getDate()}-{created.getMonth()}-{created.getFullYear()}
        </div>
      </div>
      <p className={css({ margin: '0 0 1rem' })}>{feedback}</p>
    </div>
  );
};
