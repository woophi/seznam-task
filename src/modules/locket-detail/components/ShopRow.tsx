import { faExternalLinkAlt } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ShopInfo } from 'core/models';
import { actionIcon, firstUppercase } from 'core/shared/styles';
import { FC } from 'react';
import { useFela } from 'react-fela';
type Props = {
  info: ShopInfo;
};

export const ShopRow: FC<Props> = ({ info: { name, price, link } }) => {
  const { css } = useFela();
  return (
    <a
      href={link}
      target="__blank"
      className={css(actionIcon, {
        display: 'flex',
        alignItems: 'center',
        marginTop: '.5rem',
        textDecoration: 'none',
        color: '#000',
        padding: '.5rem',
        border: '1px solid whitesmoke'
      })}
    >
      <div className={css(firstUppercase, { marginRight: '.5rem' })}>
        <b>{name}</b> - {price}$
      </div>
      <FontAwesomeIcon icon={faExternalLinkAlt} />
    </a>
  );
};
