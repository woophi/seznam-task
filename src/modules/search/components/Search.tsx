import { faFilter, faSearch } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { memo, PropsWithChildren, useState } from 'react';
import { useFela } from 'react-fela';
import { filterIcon, searchContainer, searchInput } from './styles';

type Props = PropsWithChildren<{
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}>;

export const Search = memo<Props>(({ onChange, children }) => {
  const [open, openFilters] = useState(false);
  const { css } = useFela({ open });
  return (
    <div className={css(searchContainer)}>
      <div className={css({ display: 'flex' })}>
        <FontAwesomeIcon icon={faFilter} className={css(filterIcon)} onClick={() => openFilters(v => !v)} />
        <FontAwesomeIcon icon={faSearch} className={css({ margin: '1rem 1rem 1rem .5rem' })} />
        <input onChange={onChange} className={css(searchInput)} />
      </div>
      {children}
    </div>
  );
});
