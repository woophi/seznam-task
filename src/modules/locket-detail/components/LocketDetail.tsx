import { faTimes } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { LocketDetailParameters, LocketDetailTabs } from 'core/models';
import { actionIcon, firstLetterUp } from 'core/shared/styles';
import { Tabs } from 'modules/tabs';
import { FC, useContext, useEffect } from 'react';
import { useFela } from 'react-fela';
import { LocketDetailContext } from '../context';
import { LocketParameters } from './LocketParameters';
import { ReviewRow } from './ReviewRow';
import { ShopRow } from './ShopRow';
import { detailContainer, detailImg, detailTitle } from './styles';

export const LocketDetail: FC = () => {
  const { selectedLocket, closeDetail } = useContext(LocketDetailContext);
  const { css } = useFela();

  useEffect(() => window.scrollTo({ top: 0, behavior: 'smooth' }), []);

  const parameters = selectedLocket?.parameters ?? ({} as LocketDetailParameters);

  return (
    <div className={css(detailContainer)}>
      <div className={css(detailTitle)}>
        <h1 className={css(firstLetterUp)}>{selectedLocket?.title}</h1>
        <FontAwesomeIcon icon={faTimes} onClick={closeDetail} className={css(actionIcon)} />
      </div>
      <div className={css({ display: 'flex', flexWrap: 'wrap' })}>
        <img src={selectedLocket?.imgUrl} alt={selectedLocket?.title} className={css(detailImg)} />
        <LocketParameters parameters={parameters} />
      </div>

      <p>{selectedLocket?.description}</p>
      <Tabs<LocketDetailTabs>
        tabs={[
          {
            title: 'Available in shops',
            content: selectedLocket?.shops
              .sort((a, b) => Number(a.price) - Number(b.price))
              .map(shop => <ShopRow key={shop.id} info={shop} />),
            tabId: LocketDetailTabs.Shops
          },
          {
            title: `Reviews (${selectedLocket?.reviewsCount})`,
            content: selectedLocket?.reviews
              .sort((a, b) => b.created.getTime() - a.created.getTime())
              .map(review => <ReviewRow key={review.id} info={review} />),
            tabId: LocketDetailTabs.Reviews
          }
        ]}
      />
    </div>
  );
};
