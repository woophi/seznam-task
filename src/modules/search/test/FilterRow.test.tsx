import { fireEvent, render } from '@testing-library/react';
import { noop } from 'core/data/utils';
import { FelaProvider } from 'core/fela/Provider';
import renderer from 'react-test-renderer';
import { spy } from 'sinon';
import { FilterRow } from '../components/FilterRow';

const testOptions = [
  {
    label: 'test option 1',
    value: 'test_value_1',
    checked: false
  },
  {
    label: 'test option 2',
    value: 'test_value_2',
    checked: true
  }
];

test('FilterRow renders correctly', () => {
  const component = renderer.create(
    <FelaProvider>
      <FilterRow<string> title="Test filter" options={testOptions} onChange={noop} />
    </FelaProvider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('<FilterRow />', () => {
  it('should have all required props: title, options, onChange', () => {
    const handleChange = spy();
    const { queryByText, getByTestId } = render(
      <FelaProvider>
        <FilterRow<string> title="Test Filter" options={testOptions} onChange={handleChange} />
      </FelaProvider>
    );

    expect(queryByText('Test Filter')).toBeTruthy();
    expect(queryByText(testOptions[0].label)).toBeTruthy();
    expect(queryByText(testOptions[1].label)).toBeTruthy();

    fireEvent.click(getByTestId(testOptions[0].value));
    expect(handleChange.callCount).toEqual(1);
  });
});
