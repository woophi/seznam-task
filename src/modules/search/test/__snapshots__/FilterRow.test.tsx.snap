// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterRow renders correctly 1`] = `
<div
  className="f_a f_b f_c f_d"
>
  <b
    className="f_e"
  >
    Test filter
  </b>
  <span
    className="f_e"
    data-testid="test_value_1"
    onClick={[Function]}
  >
    <span
      className="f_f"
      onClick={[Function]}
    >
      <input
        checked={false}
        className="inp-cbx"
        id="test_value_1"
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="checkbox"
      />
      <label
        className="cbx"
        htmlFor="test_value_1"
      >
        <span>
          <svg
            height="12px"
            viewBox="0 0 12 12"
            width="12px"
          >
            <polyline
              points="1 5 4 8 11 1"
            />
          </svg>
        </span>
      </label>
    </span>
    test option 1
  </span>
  <span
    className="f_e"
    data-testid="test_value_2"
    onClick={[Function]}
  >
    <span
      className="f_f"
      onClick={[Function]}
    >
      <input
        checked={true}
        className="inp-cbx"
        id="test_value_2"
        onChange={[Function]}
        style={
          Object {
            "display": "none",
          }
        }
        type="checkbox"
      />
      <label
        className="cbx"
        htmlFor="test_value_2"
      >
        <span>
          <svg
            height="12px"
            viewBox="0 0 12 12"
            width="12px"
          >
            <polyline
              points="1 5 4 8 11 1"
            />
          </svg>
        </span>
      </label>
    </span>
    test option 2
  </span>
</div>
`;
